pypsa/linopt.py

```diff
@@ -803,3 +803,74 @@ def run_and_read_gurobi(n, problem_fn, solution_fn, solver_logfile,
     del m
     return (status, termination_condition, variables_sol,
             constraints_dual, objective)
+
+
+def run_and_read_xpress(n, problem_fn, solution_fn, solver_logfile,
+                        solver_options, keep_files, warmstart=None,
+                        store_basis=True):
+    """
+    Solving function. Reads the linear problem file and passes it to
+    the Xpress solver. If the solution is successful it returns
+    variable solutions and constraint dual values. The xpress module
+    must be installed for using this function.
+
+    For more information on solver options:
+    https://www.fico.com/fico-xpress-optimization/docs/latest/solver/GUID-ACD7E60C-7852-36B7-A78A-CED0EA291CDD.html
+    """
+
+    import xpress
+
+    m = xpress.problem()
+
+    m.read(problem_fn)
+    m.setControl(solver_options)
+
+    if solver_logfile is not None:
+        m.setlogfile(solver_logfile)
+
+    if warmstart:
+        m.readbasis(warmstart)
+
+    m.solve()
+
+    if store_basis:
+        n.basis_fn = solution_fn.replace('.sol', '.bas')
+        try:
+            m.writebasis(n.basis_fn)
+        except:
+            logger.info('No model basis stored')
+            del n.basis_fn
+
+    termination_condition = m.getProbStatusString()
+
+    if termination_condition == 'mip_optimal' or \
+       termination_condition == 'lp_optimal':
+        status = 'ok'
+        termination_condition = 'optimal'
+    elif termination_condition == 'mip_unbounded' or \
+         termination_condition == 'mip_infeasible' or \
+         termination_condition == 'lp_unbounded' or \
+         termination_condition == 'lp_infeasible':
+        status = 'infeasible or unbounded'
+    else:
+        status = 'warning'
+
+    if termination_condition not in ["optimal"]:
+        return status, termination_condition, None, None, None
+
+    var = [str(v) for v in m.getVariable()]
+    variables_sol = pd.Series(m.getSolution(var), index=var).pipe(set_int_index)
+
+    try:
+        dual = [str(d) for d in m.getConstraint()]
+        constraints_dual = pd.Series(m.getDual(dual), index=dual).pipe(set_int_index)
+    except:
```

Comment:
Could you add the error type which is raise here?

Reply:
Good point, just pushed. It should catch xpress.SolverError when getDual() fails (due to e.g. unsuccessful solve). All other types of errors (InterfaceError, ModelError) are not caught, but are much more unlikely: m.getConstraint() always returns a list, possibly empty.